scipy/integrate/_cubature.py

```diff
@@ -0,0 +1,1016 @@
+    ...     np.array([1]),
+    ...     rule_1d,
+    ...     args=(alphas_1d,)
+    ... ); res.estimate
+     np.float64(1.2974425414002562)
+    >>> # Now calculate 3D integral, constructing 3D rule from product of 1D rules
+    >>> product_rule = Product([rule_1d, rule_1d, rule_1d])
+    >>> # New values of alpha, now f calcualtes exp(0.2*x_1 + 0.5*x_2 + 0.8*x_3)
+    >>> alphas_3d = np.array([0.2, 0.5, 0.8])
+    >>> res = cub(
+    ...     f,
+    ...     np.array([0, 0, 0]),
+    ...     np.array([1, 1, 1]),
+    ...     product_rule,
+    ...     args=(alphas_3d,)
+    ... ); res.estimate
+     np.float64(2.2002853017758057)
+    >>> # Evaluate for many values of alpha simultaneously
+    >>> alphas_many = np.array([[0.1, 0.2, 2], [0.25, 0.5, 5], [0.4, 0.8, 8]])
+    >>> res = cub(
+    ...     f,
+    ...     np.array([0, 0, 0]),
+    ...     np.array([1, 1, 1]),
+    ...     rule_3d,
+    ...     args=(alphas_many,)
+    ... ); res.estimate
+     array([1.46914007e+00, 2.20028530e+00, 3.50827109e+04])
+    """
+
+    if max_subdivisions is None:
+        max_subdivisions = np.inf
+
+    est = rule.estimate(f, a, b, args, kwargs)
+
+    try:
+        err = rule.error_estimate(f, a, b, args, kwargs)
+    except NotImplementedError:
+        raise Exception("attempting cubature with a rule that doesn't implement error \
+estimation.")
+
+    regions = [CubatureRegion(est, err, a, b)]
+    subdivisions = 0
+    success = False
+
+    while np.any(err > atol + rtol * np.abs(est)):
+        # region_k is the region with highest estimated error
+        region_k = heapq.heappop(regions)
+
+        est_k = region_k.estimate
+        err_k = region_k.error
+
+        a_k, b_k = region_k.a, region_k.b
+
+        # Subtract the estimate of the integral and its error from the current global
+        # estimates, since these will be refined in the loop over all subregions.
+        est -= est_k
+        err -= err_k
+
+        # Find all 2^ndim subregions formed by splitting region_k along each axis e.g.
+        # for 1D quadrature this splits an estimate over an interval into an estimate
+        # over two subintervals, for 3D cubature this splits an estimate over a cube
+        # into 8 subcubes.
+        #
+        # For each of the new subregions, calculate an estimate for the integral and
+        # the error there, and push these regions onto the heap for potential further
+        # subdividing.
+        for a_k_sub, b_k_sub in _subregion_coordinates(a_k, b_k):
+            est_sub = rule.estimate(f, a_k_sub, b_k_sub, args, kwargs)
+            err_sub = rule.error_estimate(f, a_k_sub, b_k_sub, args, kwargs)
+
+            est += est_sub
+            err += err_sub
+
+            new_region = CubatureRegion(est_sub, err_sub, a_k_sub, b_k_sub)
+
+            heapq.heappush(regions, new_region)
+
+        subdivisions += 1
+
+        if subdivisions >= max_subdivisions:
+            break
+    else:
+        success = True
```

Comment:
I think the `while` ... `else` is funny. I would set `success = True` originally, then put `success = False` right before the break of `max_subdivisions`.